rubi/rubi/contracts/market.py

```diff
@@ -1,829 +1,553 @@
-import hexbytes
-import logging as log
-from eth_abi import decode
-from eth_abi.codec import ABICodec
-from web3._utils.events import get_event_data
+from typing import Optional, Tuple, List
 
-from rubi.contracts.helper import networks
+from eth_typing import ChecksumAddress
+from web3 import Web3
+from web3.contract import Contract
 
-class RubiconMarket: 
-    """this class represents the RubiconMarket.sol contract and has read functionality.
+from rubi.contracts.base_contract import BaseContract
+from rubi.network import Network
+
+
+class RubiconMarket(BaseContract):
+    """This class represents the RubiconMarket.sol contract and by default has read functionality.
+    If a wallet and key are passed in instantiation then this class can also be used to write to the contract instance.
 
     :param w3: Web3 instance
     :type w3: Web3
-    :param contract: an optional contract instance, if not provided, the contract will be instantiated using the address and abi from the networks.py file given the chain id of the w3 instance
-    :type contract: Web3 object, optional
+    :param contract: Contract instance
+    :type contract: Contract
+    :param wallet: a wallet address of the signer (optional, default is None)
+    :type wallet: Optional[ChecksumAddress]
+    :param key: the private key of the signer (optional, default is None)
+    :type key: Optional[str]
     """
 
-    def __init__(self, w3, contract=None):
+    def __init__(
+        self,
+        w3: Web3,
+        contract: Contract,
```

Comment:
i like that we are keeping this pattern of allowing the user to pass in a custom contract here, this will help when we get to integrating this into the test suite as we will need the ability to pass arbitrary contract objects into these classes